<a href="https://colab.research.google.com/github/thfruchart/1nsi-2020/blob/master/Chap12/CORRECTION/EXERCICES_Trier_un_tableau.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Exercice 1

On souhaite compter le nombre d'opérations élémentaires réalisées lors du tri par sélection. 

Pour cela, on définit une fonction `nb_operation_tri_selection` qui :
* prend en argument un tableau t
* effectue le tri par sélection sur t
* renvoie le nombre d'opérations élémentaires effectuées : 
   * comparaison
   * affectation


1. Compléter le code ci-dessous pour remplir cet objectif.
2. Tester la fonction en calculant le nombre d'opérations nécessaires pour trier les tableaux suivants : 
   * [1,2,3]
   * [3,2,1]
   * [1,2,3,4,5,6,7,8,9,10,11,12]
3. Quel commentaire pouvez-vous faire? 

In [ ]:
def nb_operation_tri_selection(t):
    cpt = 0
    for i in range(len(t)-1):
        i_min = i 
        cpt += 2 # affectation dans i et affectation dans i_min
        for k in range(i+1, len(t)):
            cpt += 2 # affectation dans k et comparaison
            if t[i_min] > t[k] :
                i_min = k
                cpt += 1 #affectation dans i_min
        t[i], t[i_min] = t[i_min], t[i] 
        cpt += 2 # 1 echange = 2 affectations 
    return cpt

In [ ]:
nb_operation_tri_selection([1,2,3])

14

In [ ]:
nb_operation_tri_selection([3,2,1])

16

In [ ]:
nb_operation_tri_selection([1,2,3,4,5,6,7,8,9,10,11,12])

176

In [ ]:
nb_operation_tri_selection([12,11,10,9,8,7,6,5,4,3,2,1])

212

#### 3)
Commentaire, pour un grand tableau, le nombre d'opération est **beaucoup plus** important que pour un petit tableau. 

[1,2,3,4,5,6,7,8,9,10,11,12] est 4 fois plus long que [1,2,3]

mais le nombre d'opérations trouvé : 176 > 4*14 = 56

Le tri par sélection nécessite un assez grand nombre d'opérations même lorsque le tableau est déjà trié !

# Exercice 2

On souhaite mesurer le temps de calcul pour exécuter le tri par sélection sur des tableaux aléatoires de différentes tailles

1. Ecrire une fonction `tableau_aleatoire(n)` qui
   * prend en argument un nombre entier `n >= 1`
   * renvoie un tableau de longueur n, contenant des flottants aléatoires entre 0 et 1
2. Evaluer le temps nécessaire pour trier :
   * un tableau de taille 100
   * un tableau de taille 200
   * un tableau de taille 500
3. Peut-on dire que le temps de calcul est environ proportionnel à la taille du tableau ? 

#### petits rappels sur les expressions aléatoires en python

In [ ]:
from random import random

In [ ]:
[random(), random(),random() ]

[0.08170457283431676, 0.37580199774006506, 0.13396858388371435]

### correction de la question 1

In [ ]:
from random import random

def tableau_aleatoire(n):
    t = [ 0.0 ] * n 
    for i in range(n):
        t[i] = random()
    return t 

# deuxième possibilité : tableau défini par compréhension
def tableau_aleatoire(n):
    return [random() for i in range(n)]

In [ ]:
tableau_aleatoire(10)

[0.4904272665386802,
 0.9019566605843873,
 0.772002363084121,
 0.5135127104100127,
 0.9558978396566568,
 0.6247219421073265,
 0.5224886777882429,
 0.3918710998277368,
 0.43234810961488135,
 0.38182487489870043]

#### Aide pour la question 2 

Pour mesurer le temps d'exécution d'une instruction, on peut utiliser le module time, qui contient une fonction perf_counter().

Cette fonction renvoie un 'compteur de performance' aussi précis que le permet la machine, mesuré en secondes.

In [ ]:
from time import perf_counter

In [ ]:
perf_counter()

161.762296148

In [ ]:
perf_counter()

179.40135007

Pour mesurer le temps d'exécution d'une instruction, on pourra structurer son code de cette manière : 

In [ ]:
from time import perf_counter
start = perf_counter() # lancement du chrono

### début des instructions à chronométrer ###
for i in range(1000):
    pass


###  fin  des instructions à chronométrer ###
stop = perf_counter() # arrêt du chrono
print(stop - start) # affichage de la durée en seconde
# ou  affichage de la durée en milliseconde avec :  print((stop - start)*1000)


0.00011304399998834924


### correction de la question 2

on utilise la fonction tri_par_selection vue en cours

In [5]:
def tri_par_selection(t):
    for i in range(len(t)):
        i_min = i # recherche de l'indice du minimum du tableau t[i:len(t)]
        for k in range(i, len(t)):
            if t[i_min] > t[k] :
                i_min = k
        t[i], t[i_min] = t[i_min], t[i] # échange des valeurs d'indice i et i_min dans le tableau t
    return t

#### tri d'un tableau aléatoire de taille 100 :  environ 0.6 ms


In [ ]:
t = tableau_aleatoire(100)
start = perf_counter()
tri_par_selection(t)
stop = perf_counter()
print(stop - start)

0.0006430700000237266


#### tri d'un tableau aléatoire de taille 200 : environ 2.3 ms

In [ ]:
t = tableau_aleatoire(200)
start = perf_counter()
tri_par_selection(t)
stop = perf_counter()
print(stop - start)

0.002325112999983503


#### tri d'un tableau aléatoire de taille 400  : environ 9 ms

In [ ]:
t = tableau_aleatoire(400)
start = perf_counter()
tri_par_selection(t)
stop = perf_counter()
print(stop - start)

0.008910019000040847


#### tri d'un tableau aléatoire de taille 800 : environ 25 ms

In [ ]:
t = tableau_aleatoire(800)
start = perf_counter()
tri_par_selection(t)
stop = perf_counter()
print(stop - start)

0.025204527999903803


### correction de la question 3
Le temps d'exécution n'est **pas proportionnel** à la taille du tableau : il augmente plus vite que la taille du tableau!

# Exercice 3

On considère un tableau d'entiers `tab` (type `list`) dont les éléments sont des **0** ou des **1**. 

On se propose de trier ce tableau selon l'algorithme suivant : 
* à chaque étape du tri, le tableau est constitué de trois zones consécutives, 
  * la première ne contenant que des 0, 
  * la seconde n'étant pas triée 
  * et la dernière ne contenant que des 1.
* on note `i` et `j` deux entiers tels que les indices de la zone non triée sont compris entre `i` et `j` inclus. 


```
[ Zone de 0 | Zone non triée | Zone de 1]
             i              j
```


Tant que la zone non triée n'est pas réduite à un seul élément, on regarde son premier élément:
* si cet élément vaut 0, on considère qu'il appartient désormais à la zone ne contenant que des 0 ;
*  si cet élément vaut 1, il est échangé avec le dernier élément de la zone non triée et on considère alors qu’il appartient à la zone ne contenant que des 1.


Dans tous les cas, la longueur de la zone non triée diminue de 1.

Compléter le code suivant pour réaliser cette fonction

In [ ]:
def tri(tab):
    #i est le premier indice de la zone non triee, j le dernier indice.
    #Au debut, la zone non triee est le tableau entier.
    i= ...
    j= ...
    while i != j :
        if tab[i]== 0:
            i= ...
        else :
            valeur = tab[j]
            tab[j] = ...
            ...
            j= ...
    ...


# Exercice 4

Dans cet exercice, on exploite le fait que la complexité du tri par sélection est quadratique.

* On suppose qu'il faut environ 40ms pour trier un tableau de taille 1000 en exécutant le tri par sélection sur un ordinateur. 
* estimer le temps nécessaire pour trier : 
   1. Un tableau de taille 5000
   2. Un tableau de taille 10 000
   3. Un tableau de taille 50 000
* quelle est la taille maximale d'un tableau qu'on pourrait trier sur cette machine : 
   1.  en 1 heure?
   2. en 24 heures ? 

# Exercice 5

1. Ecrire un jeu de tests pour la fonction tri_par_insertion
2. Effectuer les test avec le code fourni

In [2]:
def insere(t,v,i):
    ''' insère la valeur 'v' dans le tableau 't'
    le plus à gauche possible à partir de l'indice 'i'  '''
    assert i<len(t)
    while i>0 and v<t[i-1]:
        t[i]=t[i-1]
        i = i-1
    t[i] = v
    return t

def tri_par_insertion(t):
    ''' tri du tableau t, par ordre croissant'''
    for i in range(len(t)):
        v = t[i]
        insere(t,v,i)
    return t

In [ ]:
assert tri_par_insertion([]) == []  # tableau vide
assert tri_par_insertion([1,2,3]) == [1,2,3] # tableau déjà trié
assert tri_par_insertion([30,20,10,0]) == [0,10,20,30] # tableau en ordre décroissant
assert tri_par_insertion([0,10,20,10,20,0]) == [0,0,10,10,20,20] # tableau avec répétitions
assert tri_par_insertion([-1,-2,3,-4]) == [-4,-2,-1,3] # avec valeurs négatives et positives
assert tri_par_insertion(['D','A','C','B']) == ['A','B','C','D'] # avec du texte
print('test ok')

test ok


# Exercice 6

####**1)** Tableaux à trier

* Avant d'exécuter chaque cellule, 
   * déterminer quel sera le contenu du tableau lorsque **`n=5`**, 
* puis vérifier en exécutant la cellule

In [ ]:
n = int(input('Entrer n :'))
t1 = [ i for i in range(n)]
print(t1)

Entrer n :5
[0, 1, 2, 3, 4]


In [ ]:
n = int(input('Entrer n :'))
t2 = [2*i+1 for i in range(n)]
print(t2)

Entrer n :5
[1, 3, 5, 7, 9]


In [ ]:
n = int(input('Entrer n :'))
t3 = [1000-2*i for i in range(n)]
print(t3)

Entrer n :5
[1000, 998, 996, 994, 992]


####**2)** A quoi correspond la valeur affichée par le programme de la cellule suivante ? 

In [ ]:
from time import perf_counter

n = 100
t = [2*i+1 for i in range(n)]

start = perf_counter() # lancement du chrono
tri_par_insertion(t)
stop = perf_counter() # arrêt du chrono
print('t :', (stop - start)*1000, 'ms')


t : 0.08828099998936523 ms


ce programme affiche la durée du tri par insertion sur le tableau t

####**3)** Comparer le temps d'exécution du tri par insertion pour les tableaux `t2` et `t3`, en faisant varier la taille de `n`

Que peut-on en conclure sur l'efficacité du tri par insertion ?

In [3]:
# trier un tableau initialement rangé en ordre décroissant

from time import perf_counter

n = int(input('Entrer n :'))
t = [1000-2*i for i in range(n)]  # t3

start = perf_counter() # lancement du chrono
tri_par_insertion(t)
stop = perf_counter() # arrêt du chrono
print('t3 :', (stop - start)*1000, 'ms')

Entrer n :1000
t3 : 79.05300499999157 ms


In [4]:
# trier un tableau initialement rangé en ordre croissant

from time import perf_counter

n = int(input('Entrer n :'))
t = [1000+2*i for i in range(n)]  # t3

start = perf_counter() # lancement du chrono
tri_par_insertion(t)
stop = perf_counter() # arrêt du chrono
print('t3 :', (stop - start)*1000, 'ms')

Entrer n :1000
t3 : 0.341052999999647 ms
